### DISCUSSION TITLE
Sharing state with library component

### DISCUSSION THREAD

**Discussion by cooltolia** [pmndrs/jotai#2044] [General] [answered]
Hey all. 
First of all let me say thanks for an amazing library.

On the current project I faced the following problem.
We have separate npm package with react components. They are imported by core application and then rendered consuming some props.
But the communication is so coupled that I was thinking about moving the logic to some external service, like Jotai.

I tried to create an atom and export it from npm package. And than use it in core to set some value. 
The question I have is how to make a package to get these changes inside it's component. For now as I understand `useAtom` in core and npm package are referencing two different instances of atom.

Tried using the `Provider` component, but maybe doing something wrong.

Would appreciate any help.

**Comment by dai-shi** [comment]
> For now as I understand useAtom in core and npm package are referencing two different instances of atom.

That can happen, but not exactly sure how in your case.

For the library use case, it's highly recommended to use `store` and export a custom provider and custom hooks from your library, if you want to hide Jotai internally. 

> Tried using the Provider component, but maybe doing something wrong.

Can you reproduce it without the library? Maybe with https://csb.jotai.org

**Comment by cooltolia** [comment]
Hello @dai-shi. Thanks a lot for a such quick reply.

https://codesandbox.io/s/red-snow-ffm8xp?file=/src/App.tsx
This is very dump example I am currently trying to achieve. The main difference is that `/sharedLibrary` is meant to be an external npm package in my case. And in that scenario the following example is not working.

> or the library use case, it's highly recommended to use store and export a custom provider

I met this advices too, but code samples were for jotai v1. Could you please share some minimal code example I could get the main idea?

**Reply by dai-shi** [reply]
> And in that scenario the following example is not working.

Can you elaborate what's the expected behavior and how it's not working?

**Reply by cooltolia** [reply]
Sure. As mentioned previously `/sharedLibrary` from the example above  is an npm package in my case. I build and link it to the core app during the development process.

Expected behavior would be just the same as in example.
Actual behavior: setting new value for atom in core app component doesn't trigger value update and rerender in library component.

As I understand that's because I actually end up with atoms of the same config but two different instances.

But maybe I am wrong and it should still work and the problem somewhere else on the project level. 
If you say that it still should work, I would try to create a bare minimum project setup to reproduce an issue.

**Comment by cooltolia** [comment]
Oh, I've just realized that `store` has been introduced in v2 :)
But still would be grateful for some guidance if possible

**Reply by dai-shi** [reply]
```jsx
const store = createStore();

const countAtom = atom(0);

export const CustomProvider = ({ children }) => (
  <Provider store={store}>
    {children}
  </Provider>
);

export const useCustomHookCounter = () => {
  return useAtomValue(countAtom);
};
```

**Reply by cooltolia** [reply]
Frankly speaking with this example it is not really clear what is the purpose of providing `store`. Do I need then read something from it?

**Reply by cooltolia** [reply]
Ok, finally it worked, and turned out all I had to do is to use custom hook from the example inside core app. With that change everything is working.
But as much I'm happy as much frustrated being unable to understand why this change actually resolves the issue?

**Reply by dai-shi** [reply]
`Provider` uses React Context internally, and `useAtomValue` by default uses the store in the context.
Alternatively, you can pass the store directly to the hook.
```js
  useAtomValue(countAtom, { store })
```
In that case, you don't need `Provider`.

**Comment by cooltolia** [comment, marked as answer]
Hello @dai-shi.
Just wanted to share the result of my research and say thanks for your help.
With your answers I started better understand the library plus it helped me with reading its source code.

As a result, the problem I shared is mostly related to the way how my projects are built. 
Having two projects: core and feature package I ended up with loading two copies of jotai. Each used `useAtom` from its node_modules.
And with custom hook we actually set closure on usage of `useAtom` from one place. 

Eventually it was easier simply set webpack alias for jotai to resolve it from direct node_modules and alway have only one source of library

**Reply by dai-shi** [reply]
Thanks for sharing your findings.

I think we can add a DEV-only check to detect if there are two copies in memory. 🤔

**Reply by cooltolia** [reply]
If you find it valuable. The react has this kind of check with throwing an error about wrong usage of hooks.

**Reply by dai-shi** [reply]
If I create a PR, can you check if that can help your (wrongly configured) case?
You will be able to npm install the PR's build.

**Reply by cooltolia** [reply]
Yes, sure

**Reply by dai-shi** [reply]
Here you go: #2049 

https://ci.codesandbox.io/status/pmndrs/jotai/pr/2049
☝️ Find "Local Install Instructions"

**Reply by cooltolia** [reply]
Yep, validated, warning is present:
<img width="935" alt="Screenshot 2023-07-27 at 4 17 49 PM" src="https://github.com/pmndrs/jotai/assets/20162459/a4d50d77-0867-4548-b2de-0cd553286c8a">

**Reply by ashuvssut** [reply]
I have only one version (one copy) of joti in my entire project but still I am receiving that error
Screenshot:-
- <img width="1617" alt="image" src="https://github.com/user-attachments/assets/23e76001-16eb-44cc-b72c-bee57d4efe00" />

**Comment by KDederichs** [comment]
I'm getting that warning as well now... though I only have ONE provider at the very top of my NextJS app.
Could it be that this persists between hot reloads and then triggers the warning?

**Reply by dai-shi** [reply]
Oh, good point. Though, I'm not sure if the library part is hot reloaded.
It doesn't warn on the initial load, and only warns on hot reloads? (every time?)

**Reply by itsjoeoui** [reply]
Yepp I can confirm as well, also NextJS with app router, same problem. Warning happens after every hot reload but not on initial load

**Reply by dai-shi** [reply]
The warning is still valid if you are using the default store, but NextJS users should use Provider, and you probably don't use the default store, so, it's not necessary. Let me draft a PR.

**Reply by dai-shi** [reply]
Can you guys try #2061?
https://ci.codesandbox.io/status/pmndrs/jotai/pr/2061
☝️ Find "Local Install Instructions"

**Comment by Caracolaracol** [comment]
> Can you guys try #2061? https://ci.codesandbox.io/status/pmndrs/jotai/pr/2061 ☝️ Find "Local Install Instructions"

I'm also having this issue with nextJS, tried the #2061 

![Captura de pantalla 2023-08-04 a la(s) 20 31 15](https://github.com/pmndrs/jotai/assets/87249022/82a3a3ac-1941-4ebb-a9f4-a55877d6960f)

**Reply by dai-shi** [reply]
With #2061, it should show this message `Detected multiple Jotai instances. It may cause unexpected behavior with the default store. https://github.com/pmndrs/jotai/discussions/2044`.
Are you sure you installed the codesandbox build?

**Reply by shinchima** [reply]
Hey @dai-shi I'm getting that exact warning on next@13.4.19 on hot reload. Is it safe to ignore? I couldn't see the fix for it this thread. I have the latest jotai v 2.4.1  (have tried #2061). Cheers

**Reply by dai-shi** [reply]
Do you use `getDefaultStore`?

**Reply by shinchima** [reply]
No, I don't

**Reply by dai-shi** [reply]
Hmmm, that's weird because the warning message only exists in the function. Or, is your question something other than this warning message?

https://github.com/pmndrs/jotai/blob/1d004b090d93b383d73ef9f40d978a8970afbf91/src/vanilla/store.ts#L798-L811

**Reply by dai-shi** [reply]
Oh, I see. I think you use the default store indirectly. You want to use `<Provider>` with nextjs.

**Reply by shinchima** [reply]
OK gotchya,
Thanks for being so responsive, much appreciated.

**Comment by KDederichs** [comment]
Version 2.3.0 seems to fix it, thanks :)

**Comment by abdulbari149** [comment]
Hi, I am having the same error, without using getDefaultStore or something. I am just creating a store with createStore and passing it as a prop to the Provider. I am using jotai with react-native

**Reply by dai-shi** [reply]
What's the error exactly?

**Comment by NubeBuster** [comment]
I dont fully understand the issue. I have a atoms.tsx where I have two atoms that I use in many places using the useAtom hook. I get this jotai notification and I dont know whether to ignore or if I should take action

**Reply by digoburigo** [reply]
One thing that I noticed if the React strict mode is off, this warning don't show for me at least

**Reply by NubeBuster** [reply]
Well, that probably solves it once and for all. We can ignore this warning as it's a side effect of React's strict mode. That makes sense.

**Comment by zecka** [comment]
I got this issue issue in a next js application that consume an npm package that also use jotai.
The npm package is linked via `yarn link` (not sure if it's related)

I add following config in next.config.js to fix issue
```tsx
const nextConfig = {
    webpack: (config, _context) => {
      config.resolve.alias['jotai'] = path.resolve(__dirname, 'node_modules/jotai')
      return config;
    },
}
```


Note that I got a similar issue with multiples instances of React (https://github.com/vercel/next.js/issues/50391)

**Comment by lveillard** [comment]
I'm facing exactly the same issue and i'm not getting the "Detected multiple Jotai instances` warning

We have this setup (pnpm + turborepo)
- app/nextjsApp1
- app/nextJsApp2
- package/JotaiEngine with atoms initialization
- package/SharedReactComponents, one shared lib of components which imports jotaiEngine as well

I want both nextjsApp to be independent, so they have their own provider and store (					<Provider store={atomsStore}>)

The sharedReactComponents are built with tsup but they are just like a dummy shared folder.
The 2 apps and also the sharedReactComponents use the atoms defined in the jotaiEngine

When copy-pasting the components from the shared React Components libs to each of the apps, the atoms do work, however, calling them from the sharedReactComponents is useless as they are empty.


SOLVED: 
I fixed it by adding jotai as a peer dependency in both the jotaiEngine and the reactSharedComponents. So each one of the nextjs apps are in charge of jotai even if the engine inits them.

**Reply by jelling** [reply]
I was also trying to share state between a library and an app and switching `jotai` to a peer dependency in the library fixed the issue. Since that's a less common usage of npm, here's an example:

In your library `package.json`, switch this:

```
  "dependencies": {
    "jotai": "^2.6.0"
  },
```
to:

```
  "peerDependencies": {
    "jotai": "^2.6.0"
  },
```

**Reply by Orpheus29** [reply]
Yes, this solved that problem for me.

**Comment by fjcalzado** [comment]
I'm also experiencing this issue as others say. 

In my case is even simpler: single project in providerless mode, but I can see the warning triggered from my unit tests execution with `jest` and `@testing-library`. Is it just a 'false positive'? Cause I don't think I have 2 `jotai` instances in this scenario.

UPDATE: I can confirm that:
- Running my project, `__NUMBER_OF_JOTAI_INSTANCES__` is `1`,
- However, running unit tests, `__NUMBER_OF_JOTAI_INSTANCES__` is `2`,

So, this might do the trick:
```
(globalThis as any).__NUMBER_OF_JOTAI_INSTANCES__ = -1;
``` 
if placed in `jest` setup file. However this is just hacking, I wonder why the store code is executed twice when running tests.

**Reply by NubeBuster** [reply]
Running that code just hides the problem. But I reckon we can all ignore the problem as I have not seen anyone complain of any side effects. It's just a warning.

**Reply by fjcalzado** [reply]
Agree, is just a hack. I was trying to understand why `jotai` believes there are 2 instances.

**Reply by dai-shi** [reply]
#2462 might have improved it.

**Comment by x0s3** [comment]
in case someone encounters this problem using `jest.isolatedModules`, the solution I found was as follows:

```js
let mockActualJotai: typeof import("jotai");
jest.mock("jotai", () => {
  if (!mockActualJotai) mockActualJotai = jest.requireActual("jotai");
  return mockActualJotai;
});
```

**Comment by assertnotnull** [comment]
This warning also shows in an Expo app in providerless mode.

**Reply by dai-shi** [reply]
That sounds bad unless it's caused by hot reload or something dev-only.

**Comment by tval2** [comment]
I get this warning every time I reload the page in my next.js project (but doesn't go away if I turn off strict mode)

**Reply by dai-shi** [reply]
Are you using `<Provider>`?

**Reply by tval2** [reply]
I don't think so? so I have a file called atoms/ids.ts:

```
import { atom } from "jotai";
export const idsAtom = atom<string[]>([]);
export const selectedIdAtom = atom<string | null>(null);
```

and then in 3 components I have something similar to this:
```
import { useAtom } from "jotai";
import { idsAtom, selectedIdAtom } from "@/atoms/ids";

...

export default function IdSelector() {
  const [ids] = useAtom(idsAtom);
  const [selectedId, setSelectedId] = useAtom(selectedIdAtom);
```

**Reply by dai-shi** [reply]
This is the warning to let you know that you forgot to use `<Provider>` in Next.js projects. The warning message doesn't specifically say that because it's not always the case.

https://github.com/pmndrs/jotai/blob/7d4da515040c0743907d283de145d7e81f0d6fef/docs/guides/nextjs.mdx?plain=1#L52-L68

**Reply by tval2** [reply]
oh interesting. If I'm using the newer app router next architecture do you know if this goes in layout.tsx:

```
export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
         <Provider>
           <AppRouterCacheProvider>{children}</AppRouterCacheProvider>
         </Provider> 
      </body>
    </html>
  );
}
```

vs. page.tsx

```
export default function Home() {
     return ( 
     <Provider> 
       <Component {...pageProps} /> 
     </Provider> 
   ) 
}
```

? or does it not matter?

**Reply by dai-shi** [reply]
It depends. `<Provider>` will separate the stores.

See also: #2688 #2692

**Reply by tval2** [reply]
Got it, based on that since I have a small application I put it in layout.tsx. For anyone else encountering that issue it look like this:

```
  return (
    <html lang="en">
      <body className={inter.className}>
        <Provider>
          <AppRouterCacheProvider>{children}</AppRouterCacheProvider>
        </Provider>
      </body>
    </html>
  );
```

**Reply by tval2** [reply]
wait I take this back. I still get `Detected multiple Jotai instances. It may cause unexpected behavior with the default store`

**Comment by tval2** [comment]
It is odd, even with the `Provider` wrapped around my entire app when I call `store.set(isProcessing, true);` in a utils function and confirm this is setting it correctly; then in my other component calling `  const isProcessing = useAtomValue(isProcessingAtom);` returns false. I'll keep digging but am likely missing something here.

**Reply by dai-shi** [reply]
Maybe, you want to open a new discussion as it's slightly different from the OP, and it seems Next.js specific.

**Reply by dbritto-dev** [reply]
@dai-shi I believe is something related to React Refresh if happens only on development using nextjs

**Comment by calvinusesyourcode** [comment]
For those who come after me!

I'm in NextJ2 `14.2.11` and wrapping with `<Provider />` from "jotai" was all I had to do.

```ts
import { Provider } from 'jotai'

export default function Home() {
  return (
    <Provider>
      <InfiniteCanvas items={items} />
    </Provider>
  )
}
```

**Reply by mahdix** [reply]
I think we need that even when not using NextJs. I had same error in my app (I'm not using Next) and it was gone when I added Provider.

**Comment by wuleninja** [comment]
Hi! First, thanks for the super amazing library :)

I'm developing a library that I use in a React Native app, and I'm importing its elements from local paths like this: `import XXX from '../../mylib'` (not just `'../mylib'` because my app is in a `/native` folder within a monorepo project, with backend code, etc).

And I've been surprised with atoms working as expected in components imported from `mylib`, but when used directly in my app, they failed to bear the expected state. 

Then I remembered I've had the same issue of some kind of state not being properly synced between my lib and my app. And that involved a lot of `react-*` or `react-native-*` libs. And for these libs, I found that the solution involved the Metro config. The solution was just to do the same for `Jotai`:

```js
    resolver: {
        [...]
        extraNodeModules: {
            react: path.resolve(__dirname, 'node_modules/react'),
            'react-native': path.resolve(__dirname, 'node_modules/react-native'),
            'react-native-windows': path.resolve(__dirname, 'node_modules/react-native-windows'),
            '@react-navigation/native': path.resolve(__dirname, 'node_modules/@react-navigation/native'),
            'react-native-gesture-handler': path.resolve(__dirname, 'node_modules/react-native-gesture-handler'),
            'react-native-safe-area-context': path.resolve(__dirname, 'node_modules/react-native-safe-area-context'),
            'react-native-screens': path.resolve(__dirname, 'node_modules/react-native-screens'),
            'react-native-css-interop': path.resolve(__dirname, 'node_modules/react-native-css-interop'),
            'react-native-svg': path.resolve(__dirname, 'node_modules/react-native-svg'),
            jotai: path.resolve(__dirname, 'node_modules/jotai'),
        },
```

 I guess there's an equivalent solution involving `Vite` in the React web world.

**Reply by wuleninja** [reply]
Hi everyone! Just to tell you that I've also had another issue with internal Jotai imports ("requires") not working in React Native, probably because of how Metro handles exports. A `require("jotai/react")` failed because it was expecting a `jotai/react/index.ts|js` to exist, when in reality there's a `jotai/react.js` file. So this is the solution that worked for me:

```js
// To be able to work with my ../../local_lib
const originalResolveRequest = MetroSymlinksResolver();

const config = mergeConfig(getDefaultConfig(__dirname), {
    //
    resolver: {
        // forcing Metro to find the files, instead of looking for directories
        resolveRequest: (context, moduleName, platform) => {
            if (moduleName.startsWith('jotai/')) {
                // for example: jotai/react, or jotai/vanilla
                return {
                    filePath: path.resolve(__dirname, `node_modules/${moduleName}.js`),
                    type: 'sourceFile',
                };
            }
            return originalResolveRequest(context, moduleName, platform);
        },
```

Hoping this can be useful to someone :)

**Comment by dai-shi** [comment]
@ashuvssut 

> I have only one version (one copy) of joti in my entire project but still I am receiving that error Screenshot:-
> 
> * <img alt="image" width="1617" src="https://private-user-images.githubusercontent.com/60546840/449350376-23e76001-16eb-44cc-b72c-bee57d4efe00.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg2NDgxMTcsIm5iZiI6MTc0ODY0NzgxNywicGF0aCI6Ii82MDU0Njg0MC80NDkzNTAzNzYtMjNlNzYwMDEtMTZlYi00NGNjLWI3MmMtYmVlNTdkNGVmZTAwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTMwVDIzMzAxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwZWEyYzEwMzQ5MWU2ZTAzMzY5ODZiYWY2Yzg5MDU5MDAxNTY5MTY1ZTg0MmQxOTU5MTdiNWU5YTBkZTYxMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.w2QQfF6CtPEw7ogUoihWdDLC-7LXQWrq8PUB-gKx2UQ">

Probably, webpack creates two separate builds causing two Jotai instances.
